### ISSUE TITLE
xmr-stak performance(CPU) compared to xmr-stak-cpu

### ISSUE THREAD

**Issue by madLyfe** [fireice-uk/xmr-stak#1098]
my server is running dual opterons 6276s and xmr-stak-cpu on ubuntu server. i average above 1KH. i decided to update to xmr-stak because i wanted to put a couple GPUs on the server board as well. i installed the AMD blockchain drivers and the SDK then git cloned xmr-stak and disabled CUDA. the GPUs have not been installed yet so i figured i would test out the CPU mining performance and i cant get it to the same performance level with the auto config or the same config i was using with xmr-stak-cpu.
xmr-stak-cpu 1KH+ avg:
[config(xmr-stak-cpu config).txt](https://github.com/fireice-uk/xmr-stak/files/1737629/config.xmr-stak-cpu.config.txt)
xmr-stak auto config tops out at 900h:
[cpu(xmr-stak auto config).txt](https://github.com/fireice-uk/xmr-stak/files/1737630/cpu.xmr-stak.auto.config.txt)
xmr-stak with xmr-stak-cpu config 850h avg:
[cpu(xmr-stak-cpu config on xmr-stak).txt](https://github.com/fireice-uk/xmr-stak/files/1737633/cpu.xmr-stak-cpu.config.on.xmr-stak.txt)

i can post more info upon request.

**Comment by psychocrypt** [comment]
Please compile with the dependency hwloc and remove the file cpu.txt. The
miner will generate a config with multi socket support.

**Comment by madLyfe** [comment]
so 'cmake . -DHWLOC_ENABLE=ON' ?

**Comment by psychocrypt** [comment]
yes. you config looks like you disabled it.

madLyfe <notifications@github.com> schrieb am Mo., 19. Feb. 2018, 19:13:

**Comment by madLyfe** [comment]
getting the same results

**Comment by psychocrypt** [comment]
windows or linux? Please run hwloc-ls and post the output

**Comment by madLyfe** [comment]
sorry, i updated the main post with ubuntu server.
![image](https://user-images.githubusercontent.com/16073360/36393062-a004d04a-1573-11e8-83ce-b337b8cf4785.png)

**Comment by psychocrypt** [comment]
install hwloc than you should have hwloc-ls

**Comment by madLyfe** [comment]
sorry about that. im a linux newb.
![image](https://user-images.githubusercontent.com/16073360/36393314-a787aca6-1574-11e8-9df2-0d168b8b6eef.png)

**Comment by madLyfe** [comment]
@psychocrypt is that what you were looking for via hwloc?

**Comment by psychocrypt** [comment]
Yes thats what I need. I am not sure but somehow it looks like the auto
generated config is not generated with hwloc support. IMO you should have
30 threads and 8 threads should have low power enabled.

**Comment by madLyfe** [comment]
any 8 threads of the 30? also, my autoconfig is only showing 28 threads. it leaves out 7 and 15 for 28 threads.

**Comment by Sugaroverdose** [comment]
Actually, config looks pretty legit: 
6276 Have two silicons in one package, each silicon have 2MB L2 cache per two cores and 6MB L3 inclusive cache, it's 14MB of usable cache, which is enough for 7 cryptonight threads for silicon or 14 threads per CPU.

Autosuggested config does exactly this - it's uses first 7 threads of silicon which results in 14 for each CPU

**Comment by NoleA1** [comment]
I'm in the same boat, going from xmr-stak-cpu to xmr-stak, the servers with 2x 6276 went from 1030 h/s to 930 h/s. I can give you access to one to play around with if you want.

**Comment by psychocrypt** [comment]
I did a wrong suggestion in my last post. 28 Threads should be optimal.
Never the less you can also try to remove one thread per numa socket.
Sometimes this helps.

**Comment by madLyfe** [comment]
best i can get with the auto generated file is just shy of a 1kh avg:
```
"cpu_threads_conf" :
[
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 0 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 1 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 2 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 3 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 4 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 5 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 6 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 8 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 9 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 10 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 11 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 12 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 13 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 14 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 16 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 17 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 18 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 19 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 20 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 21 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 22 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 24 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 25 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 26 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 27 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 28 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 29 },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : 30 },
],
```
best i can get is above 1kh avg with this:
```
"cpu_threads_conf" :
[
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
    { "low_power_mode" : false, "no_prefetch" : false, "affine_to_cpu" : false },
],
```
the problem with the above is sometimes it doesnt always 'stick' when i first run. i can usually get it to 'stick' above 1kh if i restart it a few times. other times it works first go.

**Comment by Sugaroverdose** [comment]
@madLyfe post hs report when it sticks to 1khs, it will give a clear picture how does os scheduler distributes threads, also i would recommend to give a try to config which i posted earlier.

**Comment by Sugaroverdose** [comment]
> i dont know what this means.
I meant screenshot of xmr-stak window after pressing 'h' to see hashrate of CPU threads
> and got significantly worse performance. was in the 500-600h/s range iirc.
That's strange, it's actually should've give not less than 700+

**Comment by jg5xd** [comment]
i have a similar issue with a dual opteron 6376.  xmrstakcpu was giving 1kH/s+ and the latest version is around 700.  largepages are enabled.  hwloc is enabled.

**Comment by jg5xd** [comment]
{
    "algo": "cryptonight",
    "background": false,
    "colors": true,
    "retries": 5,
    "retry-pause": 5,
    "donate-level": 1,
    "syslog": false,
    "log-file": null,
    "print-time": 60,
    "av": 1,
    "safe": false,
    "max-cpu-usage": 75,
    "cpu-priority": 4,
    "threads": 30,
    "pools": [
        {
            "url": "45.32.171.89:4444",
            "user": "iz5TJtrotzzcnpkqQc8bhQ4zoMd8PWnx3RHS3BkrtrYyKXXYEzAsNwmMj8XyaSjZoBZy3vwJ9BEmBVrYivWfCdDo2fH3aXAEu",
            "pass": "workerd",
            "keepalive": true,
            "nicehash": false
        }
    ],
    "api": {
        "port": 0,
        "access-token": null,
        "worker-id": null
    }
}

**Comment by jg5xd** [comment]
Alright, just thought I'd throw this in too.  I just compiled xmr-stak-cpu and it is now giving me the same hashrate as xmr-stak-(latest).  However, i have an identical machine that is still getting an average of 952 H/s.  Can't see what the difference could be.  Same Ubuntu 16.04 LTS.  Same xmr-stak install procedure.  It's just weird to have a 20%+ difference between 2 identical machines.

**Comment by DrStein99** [comment]
I have spent about 6 hours re-configuring and testing XMR_STACK on one of my systems after upgrading the processors.  I had to stop because my results were not consistent and was going nowhere.

If I launch xmr_stak, and retrieve hash-rate from HTTP API report, I write down the value.  If I restart the processes, and retrieve the hash rate again - the number is NOT the same EVERY time.  I had a 30-40 h/s variance on the same configuration on the same machine (that has total of 680 h/s). 

My next day reviewing my notes, I feel there was a possibility of the CPU frequency scaling regulator being triggered at exactly the same moment to cause a delay timing glitch somehow.  I have no real data evidence, just more theories.
